From 265919f3c122a825a715ca3f630aed038ca23851 Mon Sep 17 00:00:00 2001
From: Keerthy <j-keerthy@ti.com>
Date: Tue, 26 Aug 2014 11:13:41 +0530
Subject: [PATCH 1099/1587] rtc: OMAP: Add external 32k clock feature

Add external 32k clock feature.

Signed-off-by: Keerthy <j-keerthy@ti.com>
[zou: Original patch taken from
ti-sdk-am335x-evm-08.00.00.00-Linux-x86-Install.bin]
Signed-off-by: Cao Zou <cao.zou@windriver.com>
---
 drivers/rtc/rtc-omap.c |    9 ++++++++-
 1 files changed, 8 insertions(+), 1 deletions(-)

diff --git a/drivers/rtc/rtc-omap.c b/drivers/rtc/rtc-omap.c
index 7f77fcb..fbcbe52 100644
--- a/drivers/rtc/rtc-omap.c
+++ b/drivers/rtc/rtc-omap.c
@@ -116,6 +116,7 @@
 
 /* OMAP_RTC_OSC_REG bit fields: */
 #define OMAP_RTC_OSC_32KCLK_EN		BIT(6)
+#define OMAP_RTC_OSC_EXT_32K		BIT(3)
 
 /* OMAP_RTC_IRQWAKEEN bit fields: */
 #define OMAP_RTC_IRQWAKEEN_ALARM_WAKEEN	BIT(1)
@@ -140,6 +141,7 @@
  * the 32KHz clock to be explicitly enabled.
  */
 #define OMAP_RTC_HAS_32KCLK_EN		BIT(2)
+#define OMAP_RTC_HAS_EXT_32K		BIT(3)
 
 #define SHUTDOWN_TIME_SEC		1
 
@@ -479,7 +481,7 @@ static struct platform_device_id omap_rtc_devtype[] = {
 	[OMAP_RTC_DATA_AM3352_IDX] = {
 		.name	= "am3352-rtc",
 		.driver_data = OMAP_RTC_HAS_KICKER | OMAP_RTC_HAS_IRQWAKEEN |
-			       OMAP_RTC_HAS_32KCLK_EN,
+			       OMAP_RTC_HAS_32KCLK_EN | OMAP_RTC_HAS_EXT_32K,
 	},
 	[OMAP_RTC_DATA_DA830_IDX] = {
 		.name	= "da830-rtc",
@@ -570,6 +572,11 @@ static int __init omap_rtc_probe(struct platform_device *pdev)
 	if (id_entry->driver_data & OMAP_RTC_HAS_32KCLK_EN)
 		rtc_writel(OMAP_RTC_OSC_32KCLK_EN, OMAP_RTC_OSC_REG);
 
+	/* Enable External clock as the source */
+	if (id_entry->driver_data & OMAP_RTC_HAS_32KCLK_EN)
+		rtc_writel(OMAP_RTC_OSC_EXT_32K | rtc_read(OMAP_RTC_OSC_REG),
+			   OMAP_RTC_OSC_REG);
+
 	/* clear old status */
 	reg = rtc_read(OMAP_RTC_STATUS_REG);
 	if (!pm_off) {
-- 
1.7.5.4

